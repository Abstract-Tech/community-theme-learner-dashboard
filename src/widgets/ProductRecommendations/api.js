import { get, stringifyUrl } from 'data/services/lms/utils';
import urls from 'data/services/lms/urls';

export const getCrossProductRecommendationsUrl = (courseId) => `${urls.api}/learner_recommendations/cross_product/${courseId}`;

const fetchCrossProductCourses = (courseId) => get(stringifyUrl(getCrossProductRecommendationsUrl(courseId)));

export default {
  fetchCrossProductCourses,
};
