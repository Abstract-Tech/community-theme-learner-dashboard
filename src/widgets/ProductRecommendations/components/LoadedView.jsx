import React, { useMemo } from 'react';
import PropTypes from 'prop-types';
import { Container } from '@edx/paragon';
import { useIntl } from '@edx/frontend-platform/i18n';

import messages from '../messages';
import { courseShape, courseTypeToProductTypeMap } from '../utils';
import ProductCardContainer from './ProductCardContainer';

const MAX_OPEN_COURSE_RECOMMENDATIONS = 4;
const MIN_OPEN_COURSE_RECOMMENDATIONS = 2;

const LoadedView = ({ crossProductCourses, openCourses }) => {
  const { formatMessage } = useIntl();
  const includesCrossProductTypes = crossProductCourses.length === 2;

  const finalProductList = useMemo(() => {
    if (includesCrossProductTypes) {
      const openCourseList = openCourses ? openCourses.slice(0, MIN_OPEN_COURSE_RECOMMENDATIONS) : [];
      return crossProductCourses.concat(openCourseList);
    }
    return openCourses.slice(0, MAX_OPEN_COURSE_RECOMMENDATIONS);
  }, [crossProductCourses, openCourses, includesCrossProductTypes]);

  const courseTypes = [...new Set(finalProductList.map((item) => courseTypeToProductTypeMap[item.courseType]))];

  return (
    <div className="bg-light-200">
      <Container
        size="lg"
        className="recommendations-container pt-sm-5 pt-4.5 pb-2 pb-sm-4.5 bg-light-200"
      >
        <h2>
          {formatMessage(messages.recommendationsHeading)}
        </h2>
        <ProductCardContainer finalProductList={finalProductList} courseTypes={courseTypes} />
      </Container>
    </div>
  );
};

LoadedView.propTypes = {
  crossProductCourses: PropTypes.arrayOf(
    PropTypes.shape(courseShape),
  ).isRequired,
  openCourses: PropTypes.arrayOf(
    PropTypes.shape(courseShape),
  ).isRequired,
};

export default LoadedView;
