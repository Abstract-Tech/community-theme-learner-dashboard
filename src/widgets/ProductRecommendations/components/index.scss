
.recommendations-container {
  border: solid
}

.base-card-link:hover {
  text-decoration: none;
}