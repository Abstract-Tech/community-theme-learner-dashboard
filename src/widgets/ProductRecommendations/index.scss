@import "@edx/paragon/scss/core/core";

$horizontal-card-gap: 20px;
$vertical-card-gap: 24px;

// .recommendations-container {
//   border: solid;
// }

.base-card {
  height: 332px;
  width: 270px !important;

  p {
    margin-bottom: 0;
  }

  .pgn__card-image-cap {
    height: 104px;
    object: {
      fit: cover;
      position: top center;
    }
  }

  .pgn__card-logo-cap {
    bottom: -1.5rem;
    object: {
      fit: scale-down;
      position: center center;
    }
  }

  .product-card-title {
    font: {
      size: 1.125rem;
      weight: 700;
    }

    line-height: 24px ;
  }

  .product-card-subtitle {
    font: {
      size: 0.875rem; 
      weight: 400;
    }

    line-height: 24px;
  }

  .product-badge {
    position: absolute;
    bottom: 2.75rem;
  }

  .footer-content {
    position: absolute;
    bottom: 1rem;
  }

  &.light {
    background-color: $white;

    .title {
      color: $black;
    }

    .subtitle {
      color: $gray-700;
    }

    .badge {
      background-color: $light-500;
      color: $black;
    }

    .footer-content {
      color: $gray-700;
    }
  }

  &:hover {
    box-shadow: 0 0.125rem 0.25rem rgba(0, 0, 0, 0.15), 0 0.125rem 0.5rem rgba(0, 0, 0, 0.15);
  }
}

.base-card-link:hover {
  text-decoration: none;
}

.base-card-link .base-card {
  display: flex;
}

.product-card-container {
  gap: $vertical-card-gap $horizontal-card-gap;
  margin: 0 (-$horizontal-card-gap);
  padding: 1rem $horizontal-card-gap;

  .course-subcontainer {
    display: flex;
    gap: $vertical-card-gap $horizontal-card-gap;
  }

  @include media-breakpoint-down(lg) {
    overflow-x: scroll;
  }
}
