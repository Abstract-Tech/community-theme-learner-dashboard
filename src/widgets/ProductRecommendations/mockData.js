const mockCrossProductRecommendations = {
  courses: [
    {
      key: 'HarvardX+CRS',
      uuid: 'fee0ed87-a122-46a8-b233-3e5c75c755d1',
      title: 'CRISPR: Gene-editing Applications',
      image: {
        src: 'https://prod-discovery.edx-cdn.org/media/course/image/ed79a49b-64c1-48d2-afdc-054bf921e38d-6a76ceb47dea.small.jpg',
      },
      prospectusPath:
        'course/harvard-vpal-crispr-gene-editing-applications-online-short-course',
      owners: [
        {
          key: 'HarvardX',
          name: 'Harvard University',
          logoImageUrl:
            'http://localhost:18381/media/organization/logos/ef72daf3-c9a1-4c00-ba37-b3514392bdcf-8839c516815a.png',
        },
      ],
      activeCourseRun: {
        key: 'course-v1:HarvardX+CRS+1T2023',
        marketingUrl:
          'course/crispr-gene-editing-applications-course-v1-harvardx-crs-1t2023?utm_source=discovery_worker&utm_medium=affiliate_partner',
      },
      courseType: 'executive-education-2u',
    },
    {
      key: 'AdelaideX+BC24CYB',
      uuid: 'dacb53da-35c7-4c6b-b8d0-d7d8c0c78cef',
      title: 'Cybersecurity Boot Camp',
      image: {
        src: 'https://prod-discovery.edx-cdn.org/media/course/image/fbb62b99-9f85-4563-a67c-34ac827560bd-e3e6263b98fd.small.png',
      },
      prospectusPath:
        'course/the-university-of-adelaide-cybersecurity-boot-camp',
      owners: [
        {
          key: 'AdelaideX',
          name: 'University of Adelaide',
          logoImageUrl:
            'http://localhost:18381/media/organization/logos/51d054b4-9589-4376-9e9f-15656f3c8d0e-f10fb90c3ff1.png',
        },
      ],
      activeCourseRun: {
        key: 'course-v1:AdelaideX+BC24CYB+1T2023',
        marketingUrl:
          'course/cybersecurity-boot-camp-course-v1-adelaidex-bc24cyb-1t2023?utm_source=discovery_worker&utm_medium=affiliate_partner',
      },
      courseType: 'bootcamp-2u',
    },
    {
      key: 'AA+AA101',
      uuid: '1e2cae8c-1c67-4067-a3c0-360543e6a9b8',
      title: 'Data Analytics for Business',
      image: {
        src: 'https://prod-discovery.edx-cdn.org/media/course/image/1e2cae8c-1c67-4067-a3c0-360543e6a9b8-50beebb61f2e.small.png',
      },
      prospectusPath: '/course/data-analytics-for-business',
      owners: [
        {
          key: 'GTx',
          name: 'The Georgia Institute of Technology',
          logoImageUrl:
            'https://prod-discovery.edx-cdn.org/organization/logos/8537d31f-01b4-40fd-b652-e17b38eefe41-7956b2a3cd04.png',
        },
      ],
      activeCourseRun: {
        key: 'course-v1:GTx+MGT6203x+2T2023',
        marketingUrl:
          'https://www.edx.org/course/data-analytics-for-business-course-v1gtxmgt6203x2t2023?utm_source=prospectus_worker&utm_medium=affiliate_partner',
      },
      courseType: 'course',
    },
    {
      key: 'AA+AA101',
      uuid: '876a65e7-425b-437b-bced-bdf8059fec81',
      title: 'Western and Chinese Art: Masters and Classics',
      image: {
        src: 'https://prod-discovery.edx-cdn.org/media/course/image/876a65e7-425b-437b-bced-bdf8059fec81.small.jpg',
      },
      prospectusPath: '/course/western-and-chinese-art-masters-and-classics',
      owners: [
        {
          key: 'TsinghuaX',
          name: 'Tsinghua University',
          logoImageUrl:
            'https://prod-discovery.edx-cdn.org/organization/logos/b5714409-b5f4-4c9d-9348-b0fecbaaddd6-780fbb6c72c7.png',
        },
      ],
      activeCourseRun: {
        key: 'course-v1:TsinghuaX+00691153.x+1T2015',
        marketingUrl:
          'https://www.edx.org/course/western-chinese-art-masters-classics-tsinghuax-00691153x?utm_source=prospectus_worker&utm_medium=affiliate_partner',
      },
      courseType: 'course',
    },
  ],
};

export default mockCrossProductRecommendations;
