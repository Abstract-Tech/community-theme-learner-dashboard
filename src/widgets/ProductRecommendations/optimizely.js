import { createInstance } from '@optimizely/react-sdk';

// Remove this hardcoded value
const OPTIMIZELY_SDK_KEY = process.env.OPTIMIZELY_FULL_STACK_SDK_KEY;

const optimizelyClient = createInstance({
  sdkKey: OPTIMIZELY_SDK_KEY,
});

export default optimizelyClient;
