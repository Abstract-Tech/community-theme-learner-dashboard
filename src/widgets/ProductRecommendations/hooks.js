import { useState, useEffect } from 'react';
import { RequestStates, RequestKeys } from 'data/constants/requests';
import { StrictDict } from 'utils';
import { reduxHooks } from 'hooks';
import { SortKeys } from 'data/constants/app';
import { getCookie } from 'utils/cookies';
import { useWindowSize, breakpoints } from '@edx/paragon';
import api from './api';
import * as module from './hooks';

export const state = StrictDict({
  requestState: (val) => useState(val), // eslint-disable-line
  data: (val) => useState(val), // eslint-disable-line
});

export const useIsMobile = () => {
  const { width } = useWindowSize();
  return width < breakpoints.small.minWidth;
};

export const useShowRecommendationsFooter = () => {
  const hasAvailableDashboards = reduxHooks.useHasAvailableDashboards();
  const hasRequestCompleted = reduxHooks.useRequestIsCompleted(RequestKeys.initialize);
  const userAttributes = {
    is_mobile_user: module.useIsMobile(),
    lang_preference: global.navigator.languages.join(' '),
    is_enterprise_user: !!hasAvailableDashboards,
    location: getCookie('prod-edx-cf-loc'),
  };

  console.log("USer attributes here: ", userAttributes);

  // Hardcoded to not show until experiment related code is implemented
  return {
    shouldShowFooter: true,
    shouldLoadFooter: hasRequestCompleted && !hasAvailableDashboards,
  };
};

export const useMostRecentCourseRunKey = () => {
  const mostRecentCourseRunKey = reduxHooks.useCurrentCourseList({
    sortBy: SortKeys.enrolled,
    filters: [],
    pageSize: 0,
  }).visible[0]?.courseRun?.courseId;

  return mostRecentCourseRunKey;
};

export const useFetchRecommendations = (setRequestState, setData) => {
  const courseRunKey = module.useMostRecentCourseRunKey();

  useEffect(() => {
    let isMounted = true;

    const handleSuccess = (response) => {
      if (isMounted) {
        setData(response.data);
        setRequestState(RequestStates.completed);
      }
    };

    const handleError = () => {
      if (isMounted) {
        setRequestState(RequestStates.failed);
      }
    };

    if (courseRunKey) {
      api
        .fetchCrossProductRecommendations(courseRunKey)
        .then(handleSuccess)
        .catch(handleError);
    } else {
      api
        .fetchAmplitudeRecommendations()
        .then(handleSuccess)
        .catch(handleError);
    }
    return () => {
      isMounted = false;
    };
    /* eslint-disable */
  }, []);
};

export const useProductRecommendationsData = () => {
  const [requestState, setRequestState] = module.state.requestState(RequestStates.pending);
  const [data, setData] = module.state.data({});
  module.useFetchRecommendations(setRequestState, setData);

  return {
    productRecommendations: data,
    isLoading: requestState === RequestStates.pending,
    isLoaded: requestState === RequestStates.completed,
    hasFailed: requestState === RequestStates.failed 
  };
};

export default { useProductRecommendationsData, useShowRecommendationsFooter };
