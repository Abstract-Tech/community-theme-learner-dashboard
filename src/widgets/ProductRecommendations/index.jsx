import React from 'react';
import './index.scss';
import { useWindowSize, breakpoints } from '@edx/paragon';
import { useProductRecommendationsData } from './hooks';
import LoadingView from './components/LoadingView';
import LoadedView from './components/LoadedView';

const ProductRecommendations = () => {
  const { productRecommendations, isLoading, isLoaded } = useProductRecommendationsData();
  const { width } = useWindowSize();
  const isMobile = width < breakpoints.small.minWidth;

  if (isLoading && !isMobile) {
    return <LoadingView />;
  }

  if (isLoaded && !isMobile) {
    return (
      <LoadedView
        openCourses={productRecommendations.amplitudeCourses}
        crossProductCourses={productRecommendations.crossProductCourses}
      />
    );
  }

  return null;
};

export default ProductRecommendations;
