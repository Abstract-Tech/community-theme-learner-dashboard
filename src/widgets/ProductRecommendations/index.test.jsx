import React from 'react';
import { shallow } from 'enzyme';

import { useWindowSize } from '@edx/paragon';
import hooks from './hooks';
import ProductRecommendations from './index';
import LoadingView from './components/LoadingView';
import LoadedView from './components/LoadedView';
import { mockResponse } from './testData';

jest.mock('./hooks', () => ({
  useProductRecommendationsData: jest.fn(),
}));
jest.mock('./components/LoadingView', () => 'LoadingView');
jest.mock('./components/LoadedView', () => 'LoadedView');

describe('ProductRecommendations', () => {
  const defaultValues = {
    productRecommendations: {},
    isLoading: false,
    isLoaded: false,
    hasFailed: false,
  };

  const successfullLoadValues = {
    ...defaultValues,
    isLoaded: true,
    productRecommendations: mockResponse,
  };

  const desktopWindowSize = {
    width: 1400,
    height: 943,
  };

  it('matches snapshot', () => {
    useWindowSize.mockReturnValueOnce(desktopWindowSize);
    hooks.useProductRecommendationsData.mockReturnValueOnce({
      ...successfullLoadValues,
    });

    expect(shallow(<ProductRecommendations />)).toMatchSnapshot();
  });
  it('renders the LoadedView with course data if the request completed', () => {
    useWindowSize.mockReturnValueOnce(desktopWindowSize);
    hooks.useProductRecommendationsData.mockReturnValueOnce({
      ...successfullLoadValues,
    });

    expect(shallow(<ProductRecommendations />)).toMatchObject(
      shallow(
        <LoadedView
          openCourses={mockResponse.amplitudeCourses}
          crossProductCourses={mockResponse.crossProductCourses}
        />,
      ),
    );
  });
  it('renders the LoadingView if the request is pending', () => {
    useWindowSize.mockReturnValueOnce(desktopWindowSize);
    hooks.useProductRecommendationsData.mockReturnValueOnce({
      ...defaultValues,
      isLoading: true,
    });

    expect(shallow(<ProductRecommendations />)).toMatchObject(
      shallow(<LoadingView />),
    );
  });
  it('renders nothing if the request has failed', () => {
    useWindowSize.mockReturnValueOnce(desktopWindowSize);
    hooks.useProductRecommendationsData.mockReturnValueOnce({
      ...defaultValues,
      hasFailed: true,
    });

    const wrapper = shallow(<ProductRecommendations />);

    expect(wrapper.type()).toBeNull();
  });
  it('renders nothing if the width of the screen size is less than 576px (mobile view)', () => {
    useWindowSize.mockReturnValueOnce({ width: 575, height: 976 });
    hooks.useProductRecommendationsData.mockReturnValueOnce({
      ...successfullLoadValues,
    });

    const wrapper = shallow(<ProductRecommendations />);

    expect(wrapper.type()).toBeNull();
  });
});
