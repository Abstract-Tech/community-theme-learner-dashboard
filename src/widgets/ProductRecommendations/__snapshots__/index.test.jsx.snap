// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductRecommendations matches snapshot 1`] = `
<LoadedView
  crossProductCourses={
    Array [
      Object {
        "courseType": "executive-education-2u",
        "image": Object {
          "src": "https://www.image-2.com/ed79a49b-64c1-48d2-afdc-054bf921e38d-6a76ceb47dea.small.jpg",
        },
        "owners": Array [
          Object {
            "key": "HarvardX",
            "logoImageUrl": "http://www.image.com/ef72daf3-c9a1-4c00-ba37-b3514392bdcf-8839c516815a.png",
            "name": "Harvard University",
          },
        ],
        "prospectusPath": "course/introduction-to-computer-sceince",
        "title": "Introduction to Computer Science",
      },
      Object {
        "courseType": "bootcamp-2u",
        "image": Object {
          "src": "https://www.image-2.com/ed79a49b-64c1-48d2-afdc-054bf921e38d-6a76ceb47dea.small.jpg",
        },
        "owners": Array [
          Object {
            "key": "HarvardX",
            "logoImageUrl": "http://www.image.com/ef72daf3-c9a1-4c00-ba37-b3514392bdcf-8839c516815a.png",
            "name": "Harvard University",
          },
        ],
        "prospectusPath": "course/introduction-to-computer-sceince",
        "title": "Introduction to Computer Science",
      },
    ]
  }
  openCourses={
    Array [
      Object {
        "courseType": "verified-audit",
        "image": Object {
          "src": "https://www.image-2.com/ed79a49b-64c1-48d2-afdc-054bf921e38d-6a76ceb47dea.small.jpg",
        },
        "owners": Array [
          Object {
            "key": "HarvardX",
            "logoImageUrl": "http://www.image.com/ef72daf3-c9a1-4c00-ba37-b3514392bdcf-8839c516815a.png",
            "name": "Harvard University",
          },
        ],
        "prospectusPath": "course/introduction-to-computer-sceince",
        "title": "Introduction to Computer Science",
      },
      Object {
        "courseType": "audit",
        "image": Object {
          "src": "https://www.image-2.com/ed79a49b-64c1-48d2-afdc-054bf921e38d-6a76ceb47dea.small.jpg",
        },
        "owners": Array [
          Object {
            "key": "HarvardX",
            "logoImageUrl": "http://www.image.com/ef72daf3-c9a1-4c00-ba37-b3514392bdcf-8839c516815a.png",
            "name": "Harvard University",
          },
        ],
        "prospectusPath": "course/introduction-to-computer-sceince",
        "title": "Introduction to Computer Science",
      },
      Object {
        "courseType": "verified",
        "image": Object {
          "src": "https://www.image-2.com/ed79a49b-64c1-48d2-afdc-054bf921e38d-6a76ceb47dea.small.jpg",
        },
        "owners": Array [
          Object {
            "key": "HarvardX",
            "logoImageUrl": "http://www.image.com/ef72daf3-c9a1-4c00-ba37-b3514392bdcf-8839c516815a.png",
            "name": "Harvard University",
          },
        ],
        "prospectusPath": "course/introduction-to-computer-sceince",
        "title": "Introduction to Computer Science",
      },
      Object {
        "courseType": "course",
        "image": Object {
          "src": "https://www.image-2.com/ed79a49b-64c1-48d2-afdc-054bf921e38d-6a76ceb47dea.small.jpg",
        },
        "owners": Array [
          Object {
            "key": "HarvardX",
            "logoImageUrl": "http://www.image.com/ef72daf3-c9a1-4c00-ba37-b3514392bdcf-8839c516815a.png",
            "name": "Harvard University",
          },
        ],
        "prospectusPath": "course/introduction-to-computer-sceince",
        "title": "Introduction to Computer Science",
      },
    ]
  }
/>
`;
